src/Microsoft.Extensions.ObjectPool/DefaultObjectPool.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+
+namespace Microsoft.Extensions.ObjectPool
+{
+    public class DefaultObjectPool<T> : ObjectPool<T> where T : class, new()
+    {
+        private readonly T[] _items;
+
+        public DefaultObjectPool()
+            : this(Environment.ProcessorCount * 2)
+        {
+        }
+
+        public DefaultObjectPool(int maximumRetained)
+        {
+            _items = new T[maximumRetained];
+        }
+
+        public override T Get()
+        {
+            for (var i = 0; i < _items.Length; i++)
+            {
+                var item = _items[i];
+                if (item != null && Interlocked.CompareExchange(ref _items[i], null, item) == item)
+                {
+                    return item;
+                }
+            }
+
+            return new T();
+        }
+
+        public override void Return(T obj)
+        {
+            OnObjectReturned(obj);
+
+            for (var i = 0; i < _items.Length; i++)
```

Comment:
Returns always start at 0, so in theory if someone else is returning right before you, you're already going to look at a spot that either:
1). Already had an item
2). Just had an item returned

This aspect of the design is copied from what Roslyn does, and I don't want to change it speculatively.